// @start9labs/start-sdk/base/lib/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UseEntrypoint = exports.ExpectedExports = exports.NO_TIMEOUT = exports.SIGKILL = exports.SIGTERM = exports.inputSpecTypes = void 0;
exports.isUseEntrypoint = isUseEntrypoint;
exports.writable = writable;
exports.readonly = readonly;
exports.inputSpecTypes = __importStar(require("./actions/input/inputSpecTypes"));
__exportStar(require("./osBindings"), exports);
/** The SIGTERM signal — used for graceful daemon termination. */
exports.SIGTERM = 'SIGTERM';
/** The SIGKILL signal — used for forceful daemon termination. */
exports.SIGKILL = 'SIGKILL';
/** Sentinel value (`-1`) indicating that no timeout should be applied. */
exports.NO_TIMEOUT = -1;
/**
 * Namespace defining the required exports for a StartOS service package.
 * Every package must export implementations matching these types.
 */
var ExpectedExports;
(function (ExpectedExports) {
    version: 1;
})(ExpectedExports || (exports.ExpectedExports = ExpectedExports = {}));
/**
 * Marker class indicating that a container should use its own built-in entrypoint
 * rather than a custom command. Optionally accepts an override command array.
 */
class UseEntrypoint {
    constructor(overridCmd) {
        this.overridCmd = overridCmd;
        this.USE_ENTRYPOINT = 'USE_ENTRYPOINT';
    }
}
exports.UseEntrypoint = UseEntrypoint;
/** Type guard that checks if a {@link CommandType} is a {@link UseEntrypoint} instance. */
function isUseEntrypoint(command) {
    return typeof command === 'object' && 'USE_ENTRYPOINT' in command;
}
/** Casts a value to {@link DeepWritable} (identity at runtime, removes `readonly` at the type level). */
function writable(value) {
    return value;
}
/** Casts a value to {@link DeepReadonly} (identity at runtime, adds `readonly` at the type level). */
function readonly(value) {
    return value;
}
