// @start9labs/start-sdk/package/lib/version/VersionInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionInfo = exports.IMPOSSIBLE = void 0;
/**
 * Sentinel value indicating that a migration in a given direction is not possible.
 * Use this for `migrations.up` or `migrations.down` to prevent migration.
 */
exports.IMPOSSIBLE = Symbol('IMPOSSIBLE');
class VersionInfoImpl {
    constructor(options) {
        this.options = options;
        this._version = null;
    }
    satisfies(version) {
        return new VersionInfoImpl({
            ...this.options,
            satisfies: [...this.options.satisfies, version],
        });
    }
}
/**
 * @description Use this function to define a new version of the service. By convention, each version should receive its own file.
 * @property {string} version
 * @property {string} releaseNotes
 * @property {object} migrations
 * @returns A VersionInfo instance that is exported, then imported into versions/index.ts.
 */
exports.VersionInfo = {
    of(options) {
        return new VersionInfoImpl({ ...options, satisfies: [] });
    },
};
function __type_tests() {
    const version = exports.VersionInfo.of({
        version: '1.0.0:0',
        releaseNotes: '',
        migrations: {},
    })
        .satisfies('#other:1.0.0:0')
        .satisfies('#other:2.0.0:0')
        // @ts-expect-error
        .satisfies('#other:2.f.0:0');
    let a = version;
    // @ts-expect-error
    let b = version;
    exports.VersionInfo.of({
        // @ts-expect-error
        version: 'test',
        releaseNotes: '',
        migrations: {},
    });
    exports.VersionInfo.of({
        // @ts-expect-error
        version: 'test',
        releaseNotes: '',
        migrations: {},
    });
}
