// @start9labs/start-sdk/package/lib/test/output.sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sdk = void 0;
const StartSdk_1 = require("../StartSdk");
const setupManifest_1 = require("../manifest/setupManifest");
exports.sdk = StartSdk_1.StartSdk.of()
    .withManifest((0, setupManifest_1.setupManifest)({
    id: 'testOutput',
    title: '',
    license: '',
    packageRepo: '',
    upstreamRepo: '',
    marketingUrl: '',
    donationUrl: null,
    description: {
        short: '',
        long: '',
    },
    images: {
        main: {
            source: {
                dockerTag: 'start9/hello-world',
            },
            arch: ['aarch64', 'x86_64'],
            emulateMissingAs: 'aarch64',
        },
    },
    volumes: [],
    alerts: {
        install: null,
        update: null,
        uninstall: null,
        restore: null,
        start: null,
        stop: null,
    },
    dependencies: {
        'remote-test': {
            description: '',
            optional: false,
            s9pk: 'https://example.com/remote-test.s9pk',
        },
    },
}))
    .build(true);
