// @start9labs/start-sdk/base/lib/util/patterns.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.base64 = exports.emailWithName = exports.email = exports.domain = exports.ascii = exports.localUrl = exports.url = exports.localHostname = exports.hostname = exports.ipv4 = exports.ipv6 = void 0;
const regexes = __importStar(require("./regexes"));
/** Pattern for validating IPv6 addresses. */
exports.ipv6 = {
    regex: regexes.ipv6.matches(),
    description: 'Must be a valid IPv6 address',
};
/** Pattern for validating IPv4 addresses. */
exports.ipv4 = {
    regex: regexes.ipv4.matches(),
    description: 'Must be a valid IPv4 address',
};
/** Pattern for validating hostnames (RFC-compliant). */
exports.hostname = {
    regex: regexes.hostname.matches(),
    description: 'Must be a valid hostname',
};
/** Pattern for validating `.local` mDNS hostnames. */
exports.localHostname = {
    regex: regexes.localHostname.matches(),
    description: 'Must be a valid ".local" hostname',
};
/** Pattern for validating HTTP/HTTPS URLs. */
exports.url = {
    regex: regexes.url.matches(),
    description: 'Must be a valid URL',
};
/** Pattern for validating `.local` URLs (mDNS/LAN). */
exports.localUrl = {
    regex: regexes.localUrl.matches(),
    description: 'Must be a valid ".local" URL',
};
/** Pattern for validating ASCII-only strings (printable characters). */
exports.ascii = {
    regex: regexes.ascii.matches(),
    description: 'May only contain ASCII characters. See https://www.w3schools.com/charsets/ref_html_ascii.asp',
};
/** Pattern for validating fully qualified domain names (FQDNs). */
exports.domain = {
    regex: regexes.domain.matches(),
    description: 'Must be a valid Fully Qualified Domain Name',
};
/** Pattern for validating email addresses. */
exports.email = {
    regex: regexes.email.matches(),
    description: 'Must be a valid email address',
};
/** Pattern for validating email addresses, optionally with a display name (e.g. `"John Doe <john@example.com>"`). */
exports.emailWithName = {
    regex: regexes.emailWithName.matches(),
    description: 'Must be a valid email address, optionally with a name',
};
/** Pattern for validating base64-encoded strings. */
exports.base64 = {
    regex: regexes.base64.matches(),
    description: 'May only contain base64 characters. See https://base64.guru/learn/base64-characters',
};
